src/mjolnir/pbfgraphparser.cc

```diff
@@ -5080,16 +5104,78 @@ OSMData PBFGraphParser::ParseWays(const boost::property_tree::ptree& pt,
   for (auto& file : input_files) {
     parser.current_way_node_index_ = parser.last_node_ = parser.last_way_ = parser.last_relation_ = 0;
 
-    osmium::io::Reader reader(file, osmium::osm_entity_bits::way);
-    while (const osmium::memory::Buffer buffer = reader.read()) {
-      for (const osmium::memory::Item& item : buffer) {
-        parser.way(static_cast<const osmium::Way&>(item));
+    // These two queues maintains the order of processed ways by holding futures that correspond
+    // to the promises sent to the Lua workers. Lua workers take that promises and corresponding
+    // osmium buffers, process them and set the value of the promise.
+    using Ways = std::vector<graph_parser::Way>;
+    osmium::thread::Queue<std::future<Ways>> ways_queue(lua_concurrency * 8);
```

Comment:
Just wanna leave the note on this and other hard-coded queue and thread sizes, after our discussion earlier, that @kinkard chose them based on testing a few variations. There won't be a one-size-fits-all for all systems. For me that's fine and I'm eager to see how it'll turn out for the public instance.